redis/redis.go

```diff
@@ -82,18 +84,49 @@ type Scanner interface {
 type ConnWithTimeout interface {
 	Conn
 
-	// Do sends a command to the server and returns the received reply.
-	// The timeout overrides the read timeout set when dialing the
-	// connection.
+	// DoContext sends a command to server and returns the received reply.
+	// This function will use the timeout in this rule:
+	// If ctx doesn't have a timeout(ctx is context.Background or context.TODO),then the readtimeout which was setted when the connection is created will be used
+	// If ctx has a timeout(ctx is created by context.WithTimeout or context.WithDeadline)
+	// If ctx's timeout is bigger then the readtimeout which was setted when the connection is created,then the readtimeout will be used
+	// If ctx's timeout is smaller then the readtimeout which was setted when the connection is created,then the ctx's timeout will be used
+	DoContext(ctx context.Context, commandName string, args ...interface{}) (reply interface{}, err error)
+
+	// DoWithTimeout sends a command to the server and returns the received reply.
+	// The timeout overrides the readtimeout set when dialing the connection.
 	DoWithTimeout(timeout time.Duration, commandName string, args ...interface{}) (reply interface{}, err error)
 
-	// Receive receives a single reply from the Redis server. The timeout
-	// overrides the read timeout set when dialing the connection.
+	// ReceiveContext receives a single reply from the Redis server.
+	// This function will use the timeout in this rule:
+	// If ctx doesn't have a timeout(ctx is context.Background or context.TODO),then the readtimeout which was setted when the connection is created will be used
+	// If ctx has a timeout(ctx is created by context.WithTimeout or context.WithDeadline)
+	// If ctx's timeout is bigger then the readtimeout which was setted when the connection is created,then the readtimeout will be used
+	// If ctx's timeout is smaller then the readtimeout which was setted when the connection is created,then the ctx's timeout will be used
+	ReceiveContext(ctx context.Context) (reply interface{}, err error)
+
+	// ReceiveWithTimeout receives a single reply from the Redis server.
+	// The timeout overrides the readtimeout set when dialing the connection.
 	ReceiveWithTimeout(timeout time.Duration) (reply interface{}, err error)
 }
 
 var errTimeoutNotSupported = errors.New("redis: connection does not support ConnWithTimeout")
 
+var ErrContextCacneled = errors.New("redis: context canceled")
+
+// DoContext sends a command to server and returns the received reply.
+// This function will use the timeout in this rule:
```

Comment:
I think it would be better to talk about context generically here an example of this kind of thing can be seen here:
https://golang.org/pkg/net/http/#Server.Shutdown

In addition this should be describing the behaviour not the implementation.

Reply:
take a look at the current comment.